concrat/race-challenges/per-thread-array-join-counter-race-2.c

```diff
@@ -0,0 +1,83 @@
+// PARAM: --set ana.activated[+] symb_locks
+// Per-thread array index passed via argument.
+// Thread pool joining via threads alive counter decremented based on thread array values.
+// Extracted from smtprc.
+// TODO: symb_locks doesn't work for datas and datas_mutex here
+#include <stdlib.h>
+#include <stdbool.h>
+#include <pthread.h>
+extern void abort(void);
+void assume_abort_if_not(int cond) {
+  if(!cond) {abort();}
+}
+extern int __VERIFIER_nondet_int();
+
+int threads_total;
+int threads_alive = -1;
+pthread_mutex_t threads_alive_mutex = PTHREAD_MUTEX_INITIALIZER;
+pthread_cond_t threads_alive_cond = PTHREAD_COND_INITIALIZER;
+
+pthread_t *tids;
+bool *datas;
+pthread_mutex_t *datas_mutex;
+
+void *thread(void *arg) {
+  int i = arg;
+  pthread_mutex_lock(&datas_mutex[i]);
+  datas[i] = true; // RACE!
+  pthread_mutex_unlock(&datas_mutex[i]);
+  return NULL;
+}
+
+void *cleaner(void *arg) {
+  while (1) {
+    for (int i = 0; i < threads_total; i++) {
+      pthread_mutex_lock(&datas_mutex[i]);
+      if (datas[i]) { // NORACE
+        pthread_join(tids[i], NULL); // NORACE
+        pthread_mutex_lock(&threads_alive_mutex);
+        threads_alive--; // NORACE
+        pthread_cond_signal(&threads_alive_cond);
+        pthread_mutex_unlock(&threads_alive_mutex);
+        datas[i] = false; // NORACE
```

Comment:
So how accurate are these goblint annotations? The cleaner thread does not seem safe from a race with the unprotected access.